OpenUpSA/wazimap-ng
# Added api endpoint for dataset upload

## Description
Add api endpoint for dataset upload.

1) Added one new api endpoint to upload file to existing dataset: `/api/v1/datasets/<Dataset-ID>/upload/`
2) Added post method for `/api/datasets/` so user can create a new dataset with file object that will also create datasetfile and start upload process.


A response for updating a dataset looks something like this:
```json
{
    "id": 78,
    "created": "2021-04-28T09:57:01+0000",
    "updated": "2021-04-28T09:57:10+0000",
    "name": "Cases",
    "groups": [
        "date"
    ],
    "permission_type": "private",
    "profile": 1,
    "geography_hierarchy": 1,
    "upload_task_id": "44ceb2d8acd04343bbcb7ca8f0468b6c",
    "updated_indicators": [
        {
            "id": 74,
            "name": "Cases by month",
            "task_id": "bbd800d7dc644f1a8b095eb8f1210781"
        }
    ]
}
```
## Related Issue
#238 

## How to test it locally

1) Create/get Token 

To create ir get token user should do a post request to : `/api/v1/api-token-auth/` with fields `username and password`
It will give a response with token for user.

This token should be added to header of each request for user authentication.

User will only be able to upload dataset if group with profile name is added to user groups.
This is security where user without profile permission can not upload or create a dataset.

2) Upload to existing dataset:

to upload to existing dataset user can send post request to : `/api/v1/datasets/<Dataset-ID>/upload/`

In this case we pull profile id from Dataset.
User should have this profile group in groups to proceed upload.

file field name `file` will contain the uploaded file and header should have Authorization token 
like this : `Authorization : Token dc48cbc7f979adf7a6dcb2289a012d23462bc1a6`

3) Create a new dataset and upload file

To create a dataset send a post request to : `/api/v1/datasets/`

In this case we pull profile id from post request data as profile is required field for create dataset.
Required fields are: `profile, name and geography_hierarchy`
If user wants to upload dataset as well send file in field : `file`

Athurization token in headers is also required just like in the request above.

## Changelog

### Added

### Updated

### Removed


## Checklist


### Pull Request


### Commits


### Code Quality


### Testing
